src/families/stellar/js-synchronization.ts

```diff
@@ -1,35 +1,114 @@
+import { BigNumber } from "bignumber.js";
+import type { Account, Operation, TokenAccount } from "../../types";
 import { encodeAccountId } from "../../account";
 import type { GetAccountShape } from "../../bridge/jsHelpers";
 import { makeScanAccounts, makeSync, mergeOps } from "../../bridge/jsHelpers";
 import { fetchAccount, fetchOperations } from "./api";
+import { buildSubAccounts } from "./tokens";
 
-const getAccountShape: GetAccountShape = async (info) => {
-  const { address, initialAccount, currency, derivationMode } = info;
+const getOldAssetOperations = (
+  subAccounts?: TokenAccount[]
+):
+  | {
+      [tokenId: string]: Operation[];
+    }
+  | undefined => {
+  return subAccounts?.reduce((result, tokenAccount) => {
+    if (result[tokenAccount.id]) {
+      result[tokenAccount.id] = [
+        ...result[tokenAccount.id],
+        ...(tokenAccount.operations || []),
+      ];
+    } else {
+      result[tokenAccount.id] = tokenAccount.operations || [];
+    }
+
+    return result;
+  }, {});
+};
+
+const getAccountShape: GetAccountShape = async (info, syncConfig) => {
+  const { address, currency, initialAccount, derivationMode } = info;
   const accountId = encodeAccountId({
     type: "js",
     version: "2",
     currencyId: currency.id,
     xpubOrAddress: address,
     derivationMode,
   });
-  const oldOperations = initialAccount?.operations || [];
-  const startAt = oldOperations.length
-    ? (oldOperations[0].blockHeight || 0) + 1
-    : 0;
-  const { blockHeight, balance, spendableBalance } = await fetchAccount(
+  const { blockHeight, balance, spendableBalance, assets } = await fetchAccount(
     address
   );
-  const newOperations = await fetchOperations(accountId, address, startAt);
-  const operations = mergeOps(oldOperations, newOperations);
+
+  const oldNativeOps = initialAccount?.operations || [];
+  // Creating a map of asset operations
+  const oldAssetOps =
+    getOldAssetOperations(initialAccount?.subAccounts as TokenAccount[]) || [];
+  const oldOperations = [
+    ...oldNativeOps,
+    ...Object.values(oldAssetOps).reduce(
+      (result, ops) => [...result, ...ops],
+      []
+    ),
+  ];
+
+  const oldNativeOpsLastPagingToken = oldNativeOps[0]?.extra?.pagingToken || 0;
+  // Find the last cursor/paging token for every asset's last op
+  const oldAssetOpsLastPagingTokens = Object.values(oldAssetOps).reduce(
+    (result: string[], ops) => [...result, ops[0]?.extra?.pagingToken || 0],
+    []
+  );
+  // Find the last paging token from all
+  const lastPagingToken = BigNumber.max(
+    oldNativeOpsLastPagingToken,
+    ...oldAssetOpsLastPagingTokens,
+    0
+  ).toString();
+
+  const newOperations =
+    (await fetchOperations({
+      accountId,
+      addr: address,
+      // To fetch new records, use "desc" + "now"
+      // To fetch latest records, use "asc" + paging token
+      order: lastPagingToken === "0" ? "desc" : "asc",
+      cursor: lastPagingToken === "0" ? "now" : lastPagingToken,
```

Comment:
Wouldn't `asc + paging token` work in all situations? Unless I missed something `desc + now` returns the same operations as `asc + 0` right? With only the order being reversed, which doesn't matter if you are fetching everything.

Reply:
You're right. I missed that we're fetching the whole operations history on the initial load. Thanks for catching this!